lib/mayaUsd/fileio/translators/translatorMayaReference.cpp

```diff
@@ -424,29 +472,74 @@ MStatus UsdMayaTranslatorMayaReference::update(const UsdPrim& prim, MObject pare
-                // matches the prim we are processing.  Strip off the "_RN" suffix.
-                //
-                MString refNodeName = tempRefFn.name();
-                MString refName(refNodeName.asChar(), refNodeName.numChars() - 3);
-
-                // What reference node name is the prim expecting?
-                MString expectedRefName = refNameFromPath(parentDag);
-
-                // If found a match, reconnect the reference node's `associatedNode`
-                // attr before loading it, since the previous connection may be gone.
-                //
-                if (refName == expectedRefName) {
-                    // Reconnect the reference node's `associatedNode` attr before
-                    // loading it, since the previous connection may be gone.
-                    connectReferenceAssociatedNode(parentDag, tempRefFn);
-                    refNode = tempRefNode;
-                    break;
+        if (isMayaReference(prim)) {
+            // New behaviour for MayaReference
+            if (prim.HasAttribute(MayaReferenceAttribute)) {
+                UsdAttribute attr = prim.GetAttribute(MayaReferenceAttribute);
+                VtValue      value;
+                attr.Get(&value);
+                auto resInfo = attr.GetResolveInfo();
+                // Check if attribute was directly on prim or inherited.
+                bool isAttributeOnPrim = resInfo.GetNode().GetPath() == prim.GetPath();
+                if (isAttributeOnPrim && value.GetType() == SdfValueTypeNames->String.GetType()) {
+                    MString expectedValue = MString(value.Get<std::string>().c_str());
+                    for (MItDependencyNodes refIter(MFn::kReference); !refIter.isDone();
+                         refIter.next()) {
+                        MObject      tempRefNode = refIter.item();
+                        MFnReference tempRefFn(tempRefNode);
+                        if (!tempRefFn.isFromReferencedFile()) {
+                            if (expectedValue == tempRefFn.name()) {
+                                // Reconnect the reference node's `associatedNode` attr before
+                                // loading it, since the previous connection may be gone.
+                                connectReferenceAssociatedNode(parentDag, tempRefFn);
+                                refNode = tempRefNode;
+
+                                // Update Maya Ref Node on reconnect
+                                MString uniqueRefNodeName
+                                    = getUniqueRefNodeName(prim, parentDag, tempRefFn);
+                                tempRefFn.setName(uniqueRefNodeName);
+
+                                MString refDependNodeName = tempRefFn.name();
+                                VtValue value(std::string(
+                                    refDependNodeName.asChar(), refDependNodeName.length()));
+                                prim.SetCustomDataByKey(maya_associatedReferenceNode, value);
```

Comment:
How is this different from the "MayaReferenceAttribute"?

It seems weird to set this only when re-connecting a maya reference but not when loading it in the first place. Maybe this is old code?

Reply:
It was already in the original implementation, just updating it here in case the value changed. It has the same value as the MayaReferenceAttribute. I didn't removed it not to break any possible existing workflow using it.